File: web/src/services/taskService.ts
import { supabase } from './supabase';
import { getAuthHeaders } from './auth';
import { Task, CreateTaskInput } from '../types/tasks';

export const taskService = {
  getTasksForToday: async (): Promise<Task[]> => {
    const today = new Date().toISOString().split('T')[0];
    const { data, error } = await supabase
      .from('tasks')
      .select('*')
      .eq('date', today)
      .eq('is_archived', false)
      .order('time', { ascending: true });

    if (error) {
      console.error('[TaskService] Fetch Error:', error);
      throw error;
    }
    return data || [];
  },

  getOverdueTasks: async (): Promise<Task[]> => {
    const today = new Date().toISOString().split('T')[0];
    const { data, error } = await supabase
      .from('tasks')
      .select('*')
      .eq('status', 'pending')
      .eq('is_archived', false)
      .lt('date', today)
      .order('date', { ascending: true });

    if (error) throw error;

    // Sort: date asc (already from DB), then urgency (high→medium→low), then time
    const urgencyOrder: Record<string, number> = { high: 0, medium: 1, low: 2 };
    return (data as Task[]).sort((a, b) => {
      if (a.date !== b.date) return a.date.localeCompare(b.date);
      const uDiff = urgencyOrder[a.urgency] - urgencyOrder[b.urgency];
      if (uDiff !== 0) return uDiff;
      if (a.time && b.time) return a.time.localeCompare(b.time);
      return 0;
    });
  },

  toggleTaskStatus: async (taskId: string, currentStatus: string): Promise<void> => {
    const newStatus = currentStatus === 'completed' ? 'pending' : 'completed';
    const { error } = await supabase
      .from('tasks')
      .update({ status: newStatus })
      .eq('id', taskId);

    if (error) {
      console.error('[TaskService] Toggle Error:', error);
      throw error;
    }
  },

  createTask: async (input: CreateTaskInput): Promise<Task> => {
    const { data, error } = await supabase
      .from('tasks')
      .insert([
        {
          ...input,
          urgency: input.urgency ?? 'medium',
          status: 'pending',
          is_archived: false
        }
      ])
      .select()
      .single();

    if (error) {
      console.error('[TaskService] Create Error:', error);
      throw error;
    }
    return data;
  },

  updateTask: async (taskId: string, updates: Partial<Task>): Promise<void> => {
    const { error } = await supabase
      .from('tasks')
      .update(updates)
      .eq('id', taskId);
    if (error) throw error;
  },

  deleteTask: async (taskId: string): Promise<void> => {
    const { error } = await supabase
      .from('tasks')
      .delete()
      .eq('id', taskId);

    if (error) {
      console.error('[TaskService] Delete Error:', error);
      throw error;
    }
  },

  parseVoiceTranscript: async (transcript: string): Promise<{
    title: string | null;
    description: string | null;
    urgency: 'high' | 'medium' | 'low' | null;
    time: string | null;
  }> => {
    const headers = await getAuthHeaders();
    const backendUrl = import.meta.env.VITE_CLOUD_GATEWAY_URL || 'http://localhost:8000/api/v1';
    const response = await fetch(`${backendUrl}/tasks/parse-voice`, {
      method: 'POST',
      headers,
      body: JSON.stringify({ transcript }),
    });
    if (response.status === 503) throw new Error('warming');
    if (!response.ok) throw new Error('parse-voice failed');
    return response.json();
  },
};

File: backend/app/services/task_service.py
import os
from datetime import date
from typing import List, Dict
from supabase import create_client, Client


class TaskService:
    def __init__(self):
        url = os.environ.get("SUPABASE_URL")
        key = os.environ.get("SUPABASE_SERVICE_ROLE_KEY")
        if url and key:
            self.supabase: Client = create_client(url, key)
        else:
            self.supabase = None

    async def get_tasks(self, user_id: str, task_date: str = None, include_archived: bool = False) -> List[Dict]:
        """Fetch tasks for a given date (defaults to today)."""
        if not self.supabase:
            return []

        target_date = task_date or date.today().isoformat()
        query = self.supabase.table("tasks") \
            .select("id, date, title, description, duration, time, status, is_archived") \
            .eq("user_id", user_id) \
            .eq("date", target_date)

        if not include_archived:
            query = query.eq("is_archived", False)

        result = query.order("time", desc=False).execute()
        return result.data

    async def create_task(self, user_id: str, data: dict) -> Dict:
        """Create a new task. Defaults to today if no date provided."""
        if not self.supabase:
            return {"error": "Supabase not initialized"}

        row = {
            "user_id": user_id,
            "date": data.get("date") or date.today().isoformat(),
            "title": data["title"],
            "description": data.get("description"),
            "duration": data.get("duration"),
            "time": data.get("time"),
            "urgency": data.get("urgency", "medium"),
            "status": "pending",
            "is_archived": False,
        }

        result = self.supabase.table("tasks").insert(row).execute()
        return result.data[0] if result.data else {"error": "Insert failed"}

    async def update_task(self, user_id: str, task_id: str, updates: dict) -> Dict:
        """Update a task's fields. Only non-None fields are applied."""
        if not self.supabase:
            return {"error": "Supabase not initialized"}

        filtered = {k: v for k, v in updates.items() if v is not None}
        if not filtered:
            return {"error": "No fields to update"}

        result = self.supabase.table("tasks") \
            .update(filtered) \
            .eq("id", task_id) \
            .eq("user_id", user_id) \
            .execute()

        if not result.data:
            return {"error": "Task not found"}
        return result.data[0]

    async def delete_task(self, user_id: str, task_id: str) -> Dict:
        """Delete a task by ID."""
        if not self.supabase:
            return {"error": "Supabase not initialized"}

        result = self.supabase.table("tasks") \
            .delete() \
            .eq("id", task_id) \
            .eq("user_id", user_id) \
            .execute()

        if not result.data:
            return {"error": "Task not found"}
        return {"deleted": True, "id": task_id}
